### PR TITLE
4 - rework Buffers and Series to prepare for Timestamp support

### PR DESCRIPTION
**Pull request by antoinegelloz**
- add data type to Buffers to prepare for the timestamp support (decouple native Go type from Arrow type)

### CODE REVIEW

File: bowbuffer.go [hunk 1]
```diff
@@ -34,37 +39,34 @@ func NewBuffer(size int, typ Type) Buffer {
...
-		v[i], valid = Int64.Convert(value).(int64)
-	case []float64:
-		v[i], valid = Float64.Convert(value).(float64)
-	case []bool:
-		v[i], valid = Boolean.Convert(value).(bool)
-	case []string:
-		v[i], valid = String.Convert(value).(string)
+	switch b.DataType {
+	case Int64:
+		b.Data.([]int64)[i], valid = Int64.Convert(value).(int64)
```
Comment (yoyocm) [inline bowbuffer.go:61]: Can we consider a refactoring to remove type assertions which are increasing the technical debt and code base complexity ? 

https://medium.com/extreme-programming/avoiding-type-assertions-in-go-6feaa87

Comment (antoinegelloz) [reply bowbuffer.go:61]: I thought really hard about it already, and couldn't find a better solution. How would you do it?
I get a 404 not found on the link you sent :(

Comment (antoinegelloz) [reply bowbuffer.go:61]: Issue created

Comment (nyhu) [reply bowbuffer.go:64]: There are ways to avoid type assertion like making a dedicated buffer struct for each type and use an interface externally. Historically I choose to expose internal data to let users implement logics around their need instead of having non extensible buffer as interfaces. They are made for mutability after all. 
But after some time more methods were added like SetOrDrop etc and it might not be relevant anymore to expose internal data. Again all this have to be rethanked with generic features in mind. It's a game changer.

File: bowbuffer.go [hunk 2]
```diff
@@ -10,15 +10,20 @@ import (
 
 // Buffer is a mutable data structure with the purpose of easily building data Series with:
 // - Data: slice of data.
+// - DataType: type of the data.
 // - nullBitmapBytes: slice of bytes representing valid or null values.
 type Buffer struct {
 	Data            interface{}
+	DataType        Type
```
Comment (nyhu) [inline bowbuffer.go:17]: Historically this was here. It have been a mistake to remove it in the first place since a given type can fit several arrow types.

File: bowjson.go [hunk 3]
```diff
@@ -94,7 +95,13 @@ func (b *bow) NewValuesFromJSON(jsonB JSONBow) error {
 	*/
 
 	for fieldIndex, field := range jsonB.Schema.Fields {
-		if _, ok := mapArrowNameToBowTypes[field.Type]; ok {
+		ok := false
```
Comment (nyhu) [inline bowjson.go:98]: why switch from a map key search to a map value search, did `mapArrowNameToBowTypes` disappear ?

File: bowseries.go [hunk 4]
```diff
@@ -19,85 +19,100 @@ type Series struct {
...
+		defer builder.Release()
+		builder.AppendValues(buf.Data.([]string), buildNullBitmapBool(len(buf.Data.([]string)), buf.nullBitmapBytes))
+		return Series{Name: name, Array: builder.NewArray()}
 	default:
-		panic(fmt.Errorf("unsupported type '%T'", buf.Data))
+		panic(fmt.Errorf("unsupported type '%s'", buf.DataType))
 	}
 }
 
-func newInt64Series(name string, data []int64, valid []byte) Series {
```
Comment (nyhu) [inline bowseries.go:57]: why remove the functions since they are called twice already ? Seems like an unnecessary change ;)
Doesn't matter much, you can resolve this.

**Review by nyhu — CHANGES_REQUESTED** [review]
a single change to do.

File: bowbuffer.go (hunk 2 above)
Comment (antoinegelloz) [reply bowbuffer.go:17]: Indeed :)

File: bowbuffer.go (hunk 1 above)
Comment (antoinegelloz) [reply bowbuffer.go:64]: Yes, we need to start experimenting with the new generics, it is promising :)

File: bowjson.go (hunk 3 above)
Comment (antoinegelloz) [reply bowjson.go:98]: My mistake, I created back `mapArrowNameToBowTypes`.

File: bowseries.go (hunk 4 above)
Comment (antoinegelloz) [reply bowseries.go:57]: I just improved this ;)

**Review by yoyocm — APPROVED** [review]

**Review by nyhu — APPROVED** [review]
